### PR TITLE
Python: CWE-079 - Add Email injection query

### PR DESCRIPTION
**Pull request by mrthankyou**
This PR introduces a new query that searches for reflected XSS email injections. This query specifically looks for vulnerabilities that utilize any one of these major Python email libraries:

- [Django's Core Mail Library](https://docs.djangoproject.com/en/3.2/topics/email/)
- [Sendgrid](https://sendgrid.com/)
- [FlaskMail](https://pythonhosted.org/Flask-Mail/)
- [SmtpLib](https://docs.python.org/3/library/smtplib.html)

The most common scenario this query looks for is a Python application that utilizes an email library to deliver emails. The Python application passes user input into an email's HTML body, which can result in a reflective XSS attack on the email recipient. 

PS: Please note that both @mrthankyou and @jorgectf wrote this query. If accepted as a bug bounty we would like a 100% of the reward to go to @jorgectf. This will be re-iterated in the bug bounty submission.

### CODE REVIEW

**Comment by jorgectf** [comment]
Friendly ping @yoff :)

**Comment by yoff** [comment]
> Friendly ping @yoff :)

I was under the impression that we are still looking for a demonstrated result. But I am happy to start reviewing anyway, the code looks quite nice already :-)

**Comment by mrthankyou** [comment]
@yoff We have results, just going through the process of getting responses from teams in hopes of producing a CVE.

**Comment by jorgectf** [comment]
Made a final polish and resolved conflicts :)

**Comment by yoff** [comment]
> @yoff We have results, just going through the process of getting responses from teams in hopes of producing a CVE.

Aha, well in that case...

File: python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll
Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll:91]: This will not synchronise properly. If you do not refer to `this`, you are reporting all plaintext bodies of all SendGrid calls here. It does not really affect the query as it just defines a set of sinks as being all HTML-bodies of all calls. But is does affect the modelling and I think it will also blow up the memory usage.

If you create a concept test for `MailSender`, you should be able to see all the spurious results.

Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll:91]: This will not synchronise properly. If you do not refer to `this`, you are reporting all plaintext bodies of all SendGrid calls here. It does not really affect the query as it just defines a set of sinks as being all HTML-bodies of all calls. But is does affect the modelling and I think it will also blow up the memory usage.

File: python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll
Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll:112]: I like this construction, but again you have to synchronise on the instance. The call that `this` refers to may currently not be affected at all by the any of the calls returned by `sendgridMailHelper().getMember("Content").getACall()`.

Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll:112]: I like this construction, but again you have to synchronise on the instance. The call that `this` refer to may currently not be affected at all by the call returned by `sendgridMailHelper().getMember("Content").getACall()`.

File: python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll
Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll:126]: I appreciate the effort here. I hope we get better support for something like this in the future.

File: python/ql/src/experimental/semmle/python/Concepts.qll
Comment (yoff) [inline python/ql/src/experimental/semmle/python/Concepts.qll:425]: This seems to be the only field used in the query. Is there a reason to have and to model the rest?

**Review by yoff — CHANGES_REQUESTED** [review]
Generally good structure in the code. You will have to make the email sender concept refer to individual email senders. That means that I can write a predicate to predicate to pick out senders with certain properties, such as
```codeql
DataFlow::Node hasTestBody() {
  result.(EmailSender).getPlainTextBody().asExpression().(StrConst).getText() = "test"
}
```
This is a contrived example, but with your current implementation, this predicate will return all email senders or no email senders, depending on if one with the test body exists. This is because they all return all possible bodies for all possible senders.

File: python/ql/src/experimental/semmle/python/Concepts.qll
Comment (jorgectf) [reply python/ql/src/experimental/semmle/python/Concepts.qll:425]: https://github.com/github/codeql/pull/7127#discussion_r817897481

File: python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll
Comment (jorgectf) [reply python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll:91]: Spot on! I didn't test each sender enough. Let me find out an optimal way of correlating `this` on every predicate :)

Comment (jorgectf) [reply python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll]: As far as I have tested, this seems to correlate fine https://github.com/github/codeql/pull/7127/commits/3159d8e211a1dd874e9d2d0420a2c6764ad88709 (although it feels much more noisy 🙈)

**Comment by jorgectf** [comment]
> Generally good structure in the code. You will have to make the email sender concept refer to individual email senders. That means that I can write a predicate to predicate to pick out senders with certain properties, such as
> 
> ```ql
> DataFlow::Node hasTestBody() {
>   result.(EmailSender).getPlainTextBody().asExpression().(StrConst).getText() = "test"
> }
> ```
> 
> This is a contrived example, but with your current implementation, this predicate will return all email senders or no email senders, depending on if one with the test body exists. This is because they all return all possible bodies for all possible senders.

Having correlated the instances, this example works:

```codeql
DataFlow::Node hasTestBody() {
  result.(EmailSender).getSubject().asExpr().(StrConst).getText().matches("%Twilio%")
}
```

**Comment by mrthankyou** [comment]
@yoff This is ready for re-review. Thanks!

File: python/ql/src/experimental/semmle/python/libraries/FlaskMail.qll
Comment (jorgectf) [inline python/ql/src/experimental/semmle/python/libraries/FlaskMail.qll:70]: My bad, this also needs correlation. I will work on this soon 👍, should be easy having `getFlaskMailArgument`.

File: python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll
Comment (yoff) [reply python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll:29]: I see, I think this type tracker and the `sendgridApiSendCall` could be refactored, I will add a suggestion.

File: python/ql/src/experimental/semmle/python/libraries/FlaskMail.qll
Comment (jorgectf) [reply python/ql/src/experimental/semmle/python/libraries/FlaskMail.qll:45]: Done https://github.com/github/codeql/pull/7127/commits/930fbf777c45045840801d32d880da7709993d7f

File: python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll
Comment (jorgectf) [reply python/ql/src/experimental/semmle/python/frameworks/Sendgrid.qll]: Done https://github.com/github/codeql/pull/7127/commits/6b043446554d5a908a6f3970e34ac7b87131394e :)

**Review by yoff — DISMISSED** [review]
Thank you for the changes, I think this is now acceptable to merge into experimental. However, you might want to hold off and retain your ability to address a potential high FP rate, in case you move forward with your bounty submission?

You currently only have the string comparison sanitizer. Is that how people would normally protect themselves?

**Comment by jorgectf** [comment]
> Thank you for the changes, I think this is now acceptable to merge into experimental. However, you might want to hold off and retain your ability to address a potential high FP rate, in case you move forward with your bounty submission?
> 
> You currently only have the string comparison sanitizer. Is that how people would normally protect themselves?

Thank you for the suggestion! We've based this query's sanitizers on main's XSS query's (only `StringConstCompare` for now) https://github.com/github/codeql/blob/85f00fda1960590e833bc1c9064243d1281869d6/python/ql/lib/semmle/python/security/dataflow/SqlInjectionCustomizations.qll#L62

Would you suggest any other sanitizer? :)

**Comment by mrthankyou** [comment]
This is an interesting question about sanitizers.

I did some research and only found one sanitizer that looks to be popular called [Bleach](https://github.com/mozilla/bleach). I don't think this is something the CodeQL library accounts for at this time. It has some popularity with 2k stars and it has [163k dependent repositories](https://github.com/mozilla/bleach/network/dependents).

Bleach is pretty simple to use:

```python
import bleach

bleach.clean('evil text') // this sanitizes text
bleach.linkify('evil text with a link: github.com') // this also sanitizes text
```

It might be worth it add this sanitizer. Idk. It looks to be a pretty simple step we can add in the query. Also, this seems like something that should be baked into the CodeQL library and could deserve it's own PR.

**Comment by jorgectf** [comment]
> > Thank you for the changes, I think this is now acceptable to merge into experimental. However, you might want to hold off and retain your ability to address a potential high FP rate, in case you move forward with your bounty submission?
> > You currently only have the string comparison sanitizer. Is that how people would normally protect themselves?
> 
> Thank you for the suggestion! We've based this query's sanitizers on main's XSS query's (only `StringConstCompare` for now)
> 
> https://github.com/github/codeql/blob/85f00fda1960590e833bc1c9064243d1281869d6/python/ql/lib/semmle/python/security/dataflow/SqlInjectionCustomizations.qll#L62
> 
> Would you suggest any other sanitizer? :)

https://github.com/github/codeql/blob/85f00fda1960590e833bc1c9064243d1281869d6/python/ql/lib/semmle/python/security/dataflow/ReflectedXSSCustomizations.qll#L68

Oops, I referenced (and took into account) the wrong query type. I've added `HtmlEscaping` as sanitizer :)

**Comment by jorgectf** [comment]
> This is an interesting question about sanitizers.
> 
> I did some research and only found one sanitizer that looks to be popular called [Bleach](https://github.com/mozilla/bleach). I don't think this is something the CodeQL library accounts for at this time. It has some popularity with 2k stars and it has [163k dependent repositories](https://github.com/mozilla/bleach/network/dependents).
> 
> Bleach is pretty simple to use:
> 
> ```python
> import bleach
> 
> bleach.clean('evil text') // this sanitizes text
> bleach.linkify('evil text with a link: github.com') // this also sanitizes text
> ```
> 
> It might be worth it add this sanitizer. Idk. It looks to be a pretty simple step we can add in the query but at the same time I think Bleach also has a lot of weird ways to sanitize text which can impact how long this will take (see [here](https://bleach.readthedocs.io/en/latest/linkify.html#using-bleach-linkifier-linker)). Also, this seems like something that should be baked into the CodeQL library and could deserve it's own PR.

As discussed offline, we believe adding Bleach would require an additional PR given its complexity. We will work on that in the future in case this query arises many repos using Bleach for this :)

**Comment by yoff** [comment]
The `HtmlEscaping` is a good addition to the sanitizers. That concept can be made more powerful by adding a model of `Bleach`, but I agree that this can be done separately.

**Review by yoff — APPROVED** [review]
LGTM

**Comment by yoff** [comment]
> Considering main has the import set to public I opted to use main's branch.

It should probably be private, but I do not think you have to fix that in this PR..

File: python/ql/src/experimental/semmle/python/frameworks/Django.qll
Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Django.qll:193]:
```suggestion
    module Email {
```

**Comment by yoff** [comment]
It seems the dead code has to be removed before we can merge. I do not have permission to do it for you, but I have created suggestions that you should be able to simply accept.

**Comment by jorgectf** [comment]
Thanks @yoff, done!

**Comment by yoff** [comment]
Hm, you have to format `FlaskMail.qll` and `Sendgrid.qll`, my suggestions must have left some spaces or something...

**Comment by jorgectf** [comment]
Friendly ping @yoff :D

**Review by yoff — APPROVED** [review]
Thanks

**Pull request merged**
